File: dungeons-1.21.11/src/main/java/fishmod/features/dungeon/PuzzleDisplay.java
package fishmod.features.dungeon;

import fishmod.utils.Constants;
import fishmod.utils.config.values.FishSettings;
import config.practical.hud.HUDComponent;
import config.practical.manager.ConfigValue;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.client.network.PlayerListEntry;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class PuzzleDisplay {

    private static final List<String> puzzles = new ArrayList<>();
    private static int tickCounter = 0;

    @ConfigValue
    public static HUDComponent puzzleHud = new HUDComponent(0, 0, 150, 100, 1, "Puzzles",
            PuzzleDisplay::display,
            PuzzleDisplay::render,
            () -> true
    );

    public static void init() {
        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            if (!FishSettings.showPuzzles || client.player == null) return;
            tickCounter++;
            if (tickCounter < 20) return;
            tickCounter = 0;
            updatePuzzles(client);
        });
    }

    private static void updatePuzzles(MinecraftClient client) {
        puzzles.clear();
        ClientPlayNetworkHandler handler = client.getNetworkHandler();
        if (handler == null) return;

        Collection<PlayerListEntry> entries = handler.getPlayerList();
        String puzzleHeader = null;

        for (PlayerListEntry entry : entries) {
            if (entry.getDisplayName() == null) continue;
            String line = entry.getDisplayName().getString();
            String clean = line.replaceAll("§.", "").trim();

            if (clean.startsWith("Puzzles:")) {
                puzzleHeader = clean;
            } else if (clean.contains("[✦]") || clean.contains("[✔]") || clean.contains("???")) {
                puzzles.add(clean);
            }
        }

        if (puzzleHeader != null) {
            puzzles.add(0, puzzleHeader);
        }
    }

    public static boolean display() {
        return FishSettings.showPuzzles && !puzzles.isEmpty();
    }

    public static void render(HUDComponent component, DrawContext context) {
        MinecraftClient client = MinecraftClient.getInstance();
        int x = component.getScaledX();
        int y = component.getScaledY();

        for (int i = 0; i < puzzles.size(); i++) {
            String puzzle = puzzles.get(i);
            int color;
            if (puzzle.startsWith("Puzzles:")) color = 0xFFFFFFFF;
            else if (puzzle.contains("✔")) color = Constants.GREEN;
            else if (puzzle.contains("???")) color = Constants.BLUE;
            else color = Constants.RED;
            context.drawText(client.textRenderer, puzzle, x, y + i * Constants.TEXT_HEIGHT, color, true);
        }
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/features/dungeon/PuzzleDisplay.kt
package fishmod.features.dungeon

import config.practical.hud.HUDComponent
import config.practical.manager.ConfigValue
import fishmod.utils.Constants
import fishmod.utils.config.values.FishSettings
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.network.PlayerListEntry

object PuzzleDisplay {

    private val puzzles: MutableList<String> = ArrayList()
    private var tickCounter = 0

    @JvmField
    @ConfigValue
    var puzzleHud: HUDComponent = HUDComponent(
        0, 0, 150, 100, 1, "Puzzles",
        { display() },
        { component, context -> render(component, context) },
        { true }
    )

    @JvmStatic
    fun init() {
        ClientTickEvents.END_CLIENT_TICK.register(ClientTickEvents.EndTick { client ->
            if (!FishSettings.showPuzzles || client.player == null) return@EndTick
            tickCounter++
            if (tickCounter < 20) return@EndTick
            tickCounter = 0
            updatePuzzles(client)
        })
    }

    private fun updatePuzzles(client: MinecraftClient) {
        puzzles.clear()
        val handler = client.networkHandler ?: return

        val entries: Collection<PlayerListEntry> = handler.playerList
        var puzzleHeader: String? = null

        for (entry in entries) {
            if (entry.displayName == null) continue
            val line = entry.displayName!!.string
            val clean = line.replace(Regex("§."), "").trim()

            if (clean.startsWith("Puzzles:")) {
                puzzleHeader = clean
            } else if (clean.contains("[✦]") || clean.contains("[✔]") || clean.contains("???")) {
                puzzles.add(clean)
            }
        }

        if (puzzleHeader != null) {
            puzzles.add(0, puzzleHeader)
        }
    }

    @JvmStatic
    fun display(): Boolean {
        return FishSettings.showPuzzles && puzzles.isNotEmpty()
    }

    @JvmStatic
    fun render(component: HUDComponent, context: DrawContext) {
        val client = MinecraftClient.getInstance()
        val x = component.scaledX
        val y = component.scaledY

        for (i in puzzles.indices) {
            val puzzle = puzzles[i]
            val color: Int = if (puzzle.startsWith("Puzzles:")) 0xFFFFFFFF.toInt()
            else if (puzzle.contains("✔")) Constants.GREEN
            else if (puzzle.contains("???")) Constants.BLUE
            else Constants.RED
            context.drawText(client.textRenderer, puzzle, x, y + i * Constants.TEXT_HEIGHT, color, true)
        }
    }
}
